drivers/encx24j600/encx24j600.c

```diff
@@ -0,0 +1,424 @@
+     * */
+    reg = cmd + 0x5e;
+
+    /* read or write? bit 1 tells us */
+    if (reg & 0x2) {
+        out = ptr;
+    } else {
+        in = ptr;
+    }
+
+    /* set pointer */
+    if (addr != 0xFFFF) {
+        reg_set(dev, reg, addr);
+    }
+
+    /* copy data */
+    cmdn(dev, cmd, in, out, len);
+}
+
+static int _init(dev_eth_t *encdev)
+{
+    encx24j600_t * dev = (encx24j600_t *) encdev;
+
+    DEBUG("encx24j600: starting initialization...\n");
+
+    /* setup IO */
+    gpio_init_out(dev->cs, GPIO_PULLUP);
+    gpio_set(dev->cs);
+    gpio_init_int(dev->int_pin, GPIO_PULLUP, GPIO_FALLING, encx24j600_isr, (void*)dev);
+
+    spi_init_master(dev->spi, SPI_CONF_FIRST_RISING, ENCX24J600_SPI_SPEED);
+
+    lock(dev);
+
+    /* initialization procedure as described in data sheet (39935c.pdf) */
+    do {
+        do {
+            vtimer_usleep(ENCX24J600_INIT_DELAY);
+            reg_set(dev, EUDAST, 0x1234);
+            vtimer_usleep(ENCX24J600_INIT_DELAY);
+        } while (reg_get(dev, EUDAST) != 0x1234);
+
+        while (!(reg_get(dev, ESTAT) & CLKRDY));
+
+        /* issue System Reset */
+        cmd(dev, SETETHRST);
+
+        /* make sure initialization finalizes */
+        vtimer_usleep(1000);
+    } while (!(reg_get(dev, EUDAST) == 0x0000));
+
+    /* configure flow control */
+    phy_reg_set(dev, PHANA, 0x05E1);
+    reg_set_bits(dev, ECON2, AUTOFC);
+
+    /* setup receive buffer */
+    reg_set(dev, ERXST, RX_BUFFER_START);
+    reg_set(dev, ERXTAIL, RX_BUFFER_END);
+    dev->rx_next_ptr = RX_BUFFER_START;
+
+    /* setup interrupts */
+    reg_set_bits(dev, EIE, PKTIE | LINKIE);
+    cmd(dev, ENABLERX);
+    cmd(dev, SETEIE);
+
+    DEBUG("encx24j600: initialization complete.\n");
+
+    unlock(dev);
+
+    return 0;
+}
+
+static int _send(dev_eth_t *encdev, char *pkt, int len) {
+    encx24j600_t * dev = (encx24j600_t *) encdev;
+    lock(dev);
+
+    /* wait until previous packet has been sent */
+    while ((reg_get(dev, ECON1) & TXRTS));
+
+    /* copy packet to SRAM */
+    sram_op(dev, WGPDATA, TX_BUFFER_START, pkt, len);
+
+    /* set start of TX packet and length */
+    reg_set(dev, ETXST, TX_BUFFER_START);
+    reg_set(dev, ETXLEN, len);
+
+    /* initiate sending */
+    cmd(dev, SETTXRTS);
+
+    /* wait for sending to complete */
+    //while ((reg_get(dev, ECON1) & TXRTS));
+
+    unlock(dev);
+
+    return len;
+}
+
+/*static void encx24j600_handle_linkstate(encx24j600_t *dev, int state)
+{
+    DEBUG("encx24j600: link %s.\n", state ? "UP" : "DOWN");
+}*/
+
+static inline int packets_available(encx24j600_t *dev) {
+    /* return PKTCNT (low byte of ESTAT) */
+    return reg_get(dev, ESTAT) & ~0xFF00;
+}
+
+static void _get_mac_addr(dev_eth_t *encdev, uint8_t* buf)
+{
+    encx24j600_t * dev = (encx24j600_t *) encdev;
+    uint16_t *addr = (uint16_t *) buf;
+
+    lock(dev);
+
+    addr[0] = reg_get(dev, MAADR1);
+    addr[1] = reg_get(dev, MAADR2);
+    addr[2] = reg_get(dev, MAADR3);
```

Comment:
Not sure either. If read like this, the result is the MAC address the device responds to. If I read the bytes as coming from the register into an uint16_t, and then write them to an uint16_t array like in this function, byte-order doesn't matter, right? buf[0] will have the 1st byte, buf[5] the last?